Actors revision 2

DougGregor opened the pull request.

DougGregor committed:
Remove global actors. They will be in a separate proposal

DougGregor committed:
Defer to the separate "preventing data races" proposal for reference types.

DougGregor committed:
Allow synchronous actor functions to be accessed asynchronously.

Rework the initial narrative around this new feature, especially making
deposit(amount:) synchronous but accessed asynchronously.

DougGregor committed:
Non-actor classes cannot conform to the Actor protocol.

DougGregor committed:
Clarify the role and behavior of actor-independence

DougGregor committed:
Move discussion of closures earlier and clean it up a bit.

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -65,81 +70,190 @@ extension BankAccount {
+On the other hand, the reference to `other.accountNumber` is allowed, because `accountNumber` is immutable (defined by `let`). Once initialized, it is never written, so there can be no data races in accessing it. `accountNumber` is called *actor-independent*, because it can be freely used from any actor. Constants introduced with `let` are actor-independent by default; there is also an attribute  ...
 
-> **Note**: Constants defined by `let` are only truly immutable when the type is a value type or some kind of immutable reference type. A `let` that refers to a mutable reference type (such as a non-actor class type) would be unsafe based on the rules discussed so far. These issues are discussed later in [**Escaping reference types**](#escaping-reference-types).
+Compile-time actor-isolation checking, as shown above, ensures that code outside of the actor does not interfere with the actor's mutable state. To work with the actor from outside of it, one must always do so with an asynchronous function invocation. Asynchronous function invocations are turned into enqueues of partial tasks representing those invocations to the actor's *queue*. This queue, alon ...
```
```suggestion
Compile-time actor-isolation checking, as shown above, ensures that code outside of the actor does not interfere with the actor's mutable state. To work with the actor from outside of it, one must always do so with an asynchronous function invocation. Asynchronous function invocations are turned into enqueues of partial tasks representing those invocations to the actor's *queue* (as in, queue of tasks, not `DispatchQueue`). This queue, along with an exclusive task executor (which may be using dispatch queues) bound to the actor, functions as a synchronization boundary between the actor and any of its external callers. For example, if we wanted to make a deposit to a given bank account `account`, we could make a call to a method `deposit(amount:)`, and that call would be placed on the queue. The executor would pull tasks from the queue one-by-one, ensuring an actor is never concurrently running on multiple threads, and would eventually process the deposit.
```

Perhaps this can help clarify to people where dispatch plays into this? We always see people confused when "queue" is mentioned

DougGregor replied:
Sure, thanks. I keep wondering if we should excise the word "queue" entirely here.

ktoso replied:
We can use the good ol' "mailbox" terminology :-) It fits well here. It is the mailbox that contains tasks/messages the actor receives and has to process (on the executor).

(unresolved only to not miss that idea, feel free to take it or resolve again)

ktoso commented on proposals/nnnn-actors.md:
````diff
@@ -65,81 +70,190 @@ extension BankAccount {
+}
+```
+
+In this proposal, we assume that `forEach`'s closure parameter is *non-concurrent*, meaning that it will be executed serially, and that `parallelForEach`s closure parameter is *concurrent*, meaning that it can be executed concurrently with itself or with other code. For the purposes of this proposal, an escaping closure parameter is considered to be concurrent and a non-escaping closure parameter ...
````
Will we be able to annotate things as @nonconcurrent or something? 
If you remember the OSLog issue we hit -- the closure was escaping but synchronous.

DougGregor replied:
Yes, if escaping defaults to concurrent, we will want a @nonconcurrent annotation. I don't know how we can check it meaningfully, though

ktoso replied:
Cool yeah that'd solve it :)

DougGregor committed:
Clarify that "queue" is not a DispatchQueue

Co-authored-by: Konrad `ktoso` Malawski <ktoso@apple.com>

DougGregor committed:
Remove actor customizability via `enqueue(partialTask:)`.

This customizability will go into a separate proposal.

DougGregor committed:
Revise and expand discussion of task reentrancy, embrace non-reentrant

Revise the proposal's discussion of task reentrancy:
* Embrace non-reentrancy by default
* Add details about a new `@reentrant` attribute
* Move task-chain reentrancy into Alternatives Considered

DougGregor committed:
Discuss the removal of actor inheritance from "Alternatives Considered"

DougGregor committed:
Replace "actor class" with "actor".

DougGregor committed:
Add a Table of Contents

DougGregor committed:
Add some links for the "eliminating inheritance" discussion

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -74,7 +74,7 @@ As noted in the error message, `balance` is *actor-isolated*, meaning that it ca
 On the other hand, the reference to `other.accountNumber` is allowed, because `accountNumber` is immutable (defined by `let`). Once initialized, it is never written, so there can be no data races in accessing it. `accountNumber` is called *actor-independent*, because it can be freely used from any actor. Constants introduced with `let` are actor-independent by default; there is also an attribute  ...
 
-Compile-time actor-isolation checking, as shown above, ensures that code outside of the actor does not interfere with the actor's mutable state. To work with the actor from outside of it, one must always do so with an asynchronous function invocation. Asynchronous function invocations are turned into enqueues of partial tasks representing those invocations to the actor's *queue*. This queue, alon ...
+Compile-time actor-isolation checking, as shown above, ensures that code outside of the actor does not interfere with the actor's mutable state. To work with the actor from outside of it, one must always do so with an asynchronous function invocation. Asynchronous function invocations are turned into enqueues of partial tasks representing those invocations to the actor's *queue* (as in, queue of  ...
```
great, thanks :) This should help people grasp what we talk about

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -589,12 +589,14 @@ By default, the mutable stored properties (declared with `var`) of an actor clas
-protocol Actor: AnyObject {
-  func enqueue(partialTask: PartialAsyncTask)
-}
+protocol Actor: AnyObject { /* unspecified */ }
```
👍 Great, so we can separately talk about these once executors pitch is out there...

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -589,12 +589,14 @@ By default, the mutable stored properties (declared with `var`) of an actor clas
-The `enqueue(partialTask:)` operation is a low-level operation used to queue work for the actor to execute. `PartialAsyncTask` represents a unit of work to execute. It effectively has a single synchronous function, `run()`, which should be called synchronously within the actor's context. Only the compiler can produce new `PartialAsyncTasks`. To explicitly enqueue work on an actor, use the `run` m ...
+All actor classes implicitly conform to the `Actor` protocol. Non-actor classes cannot conform to the `Actor` protocol.
+
+The `Actor` protocol has no stated requirements in this proposal. Rather, a separate proposal on customized executors will specify the requirements and detail how to provide an alternative implementation of an actor's queueing behavior. Regardless of those details, an actor that does not provide implementations of those requirements will receive synthesized default implementations. Such actors ar ...
```
👍 

And thanks a lot for defining "default actor" I kind of guessed meaning from our impls, but wasnt defined previously :)

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -589,12 +589,14 @@ By default, the mutable stored properties (declared with `var`) of an actor clas
+
+The `Actor` protocol has no stated requirements in this proposal. Rather, a separate proposal on customized executors will specify the requirements and detail how to provide an alternative implementation of an actor's queueing behavior. Regardless of those details, an actor that does not provide implementations of those requirements will receive synthesized default implementations. Such actors ar ...
+
+The `Actor` protocol provides a single operation, the `run` method, which executes a given function on this actor:
```
Oh you actually mean to provide it built in?

I thought this was an example "if you want to, you can put such run() onto your actor".

- Do all actors really need to be thrown a random closure at them?
- The boilerplate of adding such function is not that high if we don't pre-define it, anyone who wants their actor to run random stuff can simply define such function on it.
  - they are also then in charge of it being `public` or `internal` which may matter; I would not want to allow anyone to run a closure which may block and do terrible things on my "framework internal actor" even if the actor is public... so at least not pre-defining this function would be better imo.

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -254,23 +254,19 @@ extension BankAccount {
+### Actor reentrancy
 
-One critical point that needs to be discussed and fleshed out is whether actors are [reentrant](https://en.wikipedia.org/wiki/Reentrancy_(computing)) by default or not.
+One critical point that needs to be discussed is whether actor-isolated functions are [reentrant](https://en.wikipedia.org/wiki/Reentrancy_(computing)). When an actor-isolated function suspends, reentrancy allows other work to execute on the actor before the original actor-isolated function resumes, which we refer to as *interleaving*. Reentrancy eliminates a source of deadlocks, where two actors ...
```
well phrased intro, thanks!

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -254,23 +254,19 @@ extension BankAccount {
-The notion of reentrancy allows the actor runtime to claim the complete elimination of deadlocks, offers opportunity for scheduling optimization techniques where a "high priority task" _must_ be executed as soon as possible, however at the cost of _interleaving_ with any other actor-isolated function's execution any other asynchronous function declared on this actor. This imposes a large mental b ...
+This section explores the issue of reentrancy with examples that illustrate both the benefits and problems with both reentrant and non-reentrant actors, and settles on the following overall approach:
+* Actors default to being non-reentrant, but
+* An attribute `@reentrant` allows one to make a particular actor-isolated function reentrant. 
```
Excellent, we've just saved tons of people utter confusion and endless race debugging sessions 😎 👍

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -490,55 +424,32 @@ actor class ImageDownloader {
-Thanks to Swift embracing [Structured Concurrency](https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md) as a core building block of it's concurrency story, we are in good position to do _better_ than just outright banning reentrancy. In Swift, every asynchronous operation is part of a `Task` which encapsulates the general computation  ...
-
-This resolves both the deadlock we encountered in the `Waiter` and `Kitchen` example in the previous section, and even enables implementing mutually recursive actors. In the following–world's silliest isEven implementation, we can see two actors performing either is even/odd checks, and mutually calling out to each other, because of the structured nature of tasks, and task awareness of the actor  ...
+An actor class itself can be annotated as `@reentrant`, to indicate that all actor-isolated async functions of that actor are `@reentrant` by default. The `@reentrant(never)` form of the attribute can be used to disable re-entrancy for specific functions:
```
sounds good, including the never variant 👍

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -718,6 +629,25 @@ extension BankAccount {
+* `@reentrant`: Indicates that each potential suspension point within the function body is reentrant.
+* `@reentrant(never)`: Indicates that each potential suspension point within the function body is non-reentrant.
+
+A non-reentrant potential suspension point prevents any other asynchronous call from executing on the actor until it has completed. Note that asynchronous calls to non-reentrant async functions directly on `self` are exempted from this check, so an actor can asynchronously call itself without producing a deadlock.
```
👍

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -744,3 +677,64 @@ Nearly all changes in actor isolation are breaking changes, because the actor is
+We could introduce a new kind of reentrancy, *task-chain reentrancy*, which allows reentrant calls on behalf of the given task or any of its children. This resolves both the deadlock we encountered in the `Waiter` and `Kitchen` example from the section on [deadlocks](#deadlocks-with-non-reentrant-actors) as well as the mutually-recursive `isEven` example above, while still preventing reentrancy f ...
+
+There are a few reasons why we are not currently comfortable including task-chain reentrancy in the proposal:
+* The task-based reentrancy approach doesn't seem to have been tried at scale. Orleans documents support for [reentrancy in a call chain](https://dotnet.github.io/orleans/docs/grains/reentrancy.html#reentrancy-within-a-call-chain), but the implementation is fairly limited.
```
You could even add:


What's more, it actually is _not present_ in the current Orleans runtime, as it was confusing and disapointing in real world scenarios, as mentioned by the current project's lead developer Reuben Bond [here](https://twitter.com/reubenbond/status/1349725703634251779).

In case you'd like to shut down people from trying to invent reasons that it's awesome and "orleans has it". :)

DougGregor replied:
Good point, I'll include this note, because it's important that they tried it and it didn't work out for them. That doesn't necessarily mean it's wrong for Swift, but it's a concern!

ktoso replied:
Yup, exactly. It's a bit too tempting to point at it and wave hands around "make it work, it's magic" otherwise :D

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -744,3 +677,64 @@ Nearly all changes in actor isolation are breaking changes, because the actor is
+* We do not yet know of an efficient implementation technique for this approach within the actor runtime.
+* The approach is semantically similar to recursive locks (even though there is no actual locking or blocking involved), which are particularly hard to work with in concurrent code, and we have not yet convinced ourselves those problems won't manifest in the actor model.
+
+If we can address the above, task-chain reentrancy can be introduced into the actor model with another spelling of the reentrancy attribute such as `@reentrant(task)`, and may provide a more suitable default than non-reentrant (`@reentrant(never)`).
```
ah, that's a nice hatch in case we ever solve this, cool

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -738,3 +723,27 @@ There are a few reasons why we are not currently comfortable including task-chai
+* Actor inheritance makes it easier to port existing class hierarches to get the benefits of actors. Without actor inheritance, such porting will also have to content with (e.g.) replacing superclasses with protocols and explicitly-specified stored properties.
+* The lack of inheritance in actors won't prevent users from having to understand the complexities of inheritance, because inheritance will still be used pervasively with classes.
+* The design and implementation of actors naturally admits inheritance. Actors are fundamentally class-like, the semantics of inheritance follow directly from the need to maintain actor isolation, and the implementation of actors as "special classes" naturally supports all of these features. There is little benefit to the implementation from eliminating the possibility of inheritance of actors.
+
```
That's incomplete / not concluded yet right?

I guess in pitch#2 let's hash this out 👍

DougGregor replied:
I'll add a conclusion, but yes, this is still a point of debate.

ktoso replied:
Ok 👍

ktoso commented on proposals/nnnn-actors.md:
```diff
@@ -25,12 +25,12 @@ Actors provide a model for building concurrent programs that are free of data ra
 ## Proposed solution
 
-### Actor classes
+### Actors
```
<3

ktoso commented:
Looking good! Thanks for the cleanup and I think `@reentrant` is exactly the way to go 👍 
Love the `actor` move as well; I'm not sure if we decided inheritance stuff yet, but I guess this discussion will continue :)

etcwilde commented on proposals/nnnn-actors.md:
```diff
@@ -55,112 +94,217 @@ extension BankAccount {
-It should be noted that actor isolation adds a new dimension, separate from access control, to the decision making process whether or not one is allowed to invoke a specific function on an actor. Specifically, synchronous functions may only be invoked by the specific actor instance itself, and not even by any other instance of the same actor class. 
-
-All interactions with an actor (other than the special-cased access to constants) must be performed asynchronously (semantically, one may think about this as the actor model's messaging to and from the actor). Asynchronous functions provide a mechanism that is suitable for describing such operations, and are explained in depth in the complementary [async/await proposal](https://github.com/DougGre ...
+Synchronous actor functions can be called synchronously on the actor's `self`, but must be called asynchronously the from outside of the actor. The `transfer(amount:to:)` function calls it asynchronously (on `other`), while the following function `passGo` calls it synchronously (on the implicit `self`):
```
s/asynchronously the from/asynchronously from/

DougGregor committed:
Remove the Actor protocol; it's not useful in this proposal.

The Actor protocol will come back with a separate proposal that allows
non-default actors.

DougGregor committed:
Improve the discussion of reentrancy further.

Back to reentrant by default, with yet more justification.

DougGregor committed:
Address more feedback
